### PR TITLE
eth/tracers: add golang 4byte tracer

### PR DESCRIPTION
**Pull request by wardbradt**

### CODE REVIEW

File: eth/tracers/native/4byte.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+// newFourByteTracer returns a native go tracer which collects
+// 4 byte-identifiers of a tx, and implements vm.EVMLogger.
+func newFourByteTracer() tracers.Tracer {
+	t := &fourByteTracer{
+		ids: make(map[string]int),
+	}
+	return t
+}
+
+// logic borrowed from newJsTracer in eth/tracers/js/tracer.go
```
Comment (holiman) [inline eth/tracers/native/4byte.go:64]:
```suggestion
// isPrecompiled returns whether the addr is a precompile. Logic borrowed from newJsTracer in eth/tracers/js/tracer.go
```

File: eth/tracers/native/4byte.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+		t.env.Cancel()
+		return
+	}
+	// Skip any pre-compile invocations, those are just fancy opcodes
+	if t.isPrecompiled(to) {
+		return
+	}
+	if len(input) >= 4 {
+		t.store(input[0:4], len(input)-4)
+	}
```
Comment (holiman) [inline eth/tracers/native/4byte.go:111]: Might be slightly faster if you do the quick `len(input)` check before the `isPrecompile`, which is slightly more costly. So
```suggestion
	if len(input) < 4 {
		return
	}	
	// Skip any pre-compile invocations, those are just fancy opcodes
	if t.isPrecompiled(to) {
		return
	}
	t.store(input[0:4], len(input)-4)
```

File: eth/tracers/native/4byte.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package native
+
+import (
+	"encoding/json"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/ethereum/go-ethereum/eth/tracers"
```
Comment (holiman) [inline eth/tracers/native/4byte.go:23]: needs a goimports formatting

**Review by holiman — COMMENTED** [review]
LGTM, thanks!

File: eth/tracers/native/4byte.go (hunk 3 above)
Comment (wardbradt) [reply eth/tracers/native/4byte.go]: fixed here https://github.com/ethereum/go-ethereum/pull/23882/commits/a2763283b43b248f01c98cdc2dbc7499163547a4

**Comment by holiman** [comment]
Some timing stats

native: 
```
INFO [11-11|08:30:28.658] Chain tracing finished                   start=13,401,259 end=13,402,259 transactions=188,695 elapsed=8m22.495952253s
```
js
```
INFO [11-11|08:53:48.225] Chain tracing finished                   start=13,401,259 end=13,402,259 transactions=188,695 elapsed=12m24.845998177s
```

js-legacy
```
INFO [11-11|09:48:25.003] Chain tracing finished                   start=13,401,259 end=13,402,259 transactions=188,695 elapsed=30m36.802869401s
```
I'm going to go through the results and see if there are any discrepancies

**Comment by holiman** [comment]
I've found zero discrepancies between js and go, however, there are differences between (js/go) vs legacy. They are of this type: 

```
{
    ...skipped 2 object properties...,
    "traces": [
        ...skipped 50 array elements...,
        {
            "result": {
                ...skipped 1 object property...,
                "0x60806040-1708": 1,
                ...skipped 1 object property...
            }
        },
        ...skipped 106 array elements...,
        {
            "result": {
                "0x3d602d80-51": 1,
                ...skipped 8 object properties...
            }
        },
```
Which looks like code, to me. I suspect the legacy one _correctly_ does not catch initcode, but the newer traces doesn't. So the new tracers needs to fixed. I'll push a fix

File: eth/tracers/native/4byte.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+	"sync/atomic"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/ethereum/go-ethereum/eth/tracers"
+)
+
+func init() {
+	register("4byte", newFourByteTracer)
```
Comment (s1na) [inline eth/tracers/native/4byte.go:32]: So this would introduce a new tracer. If we want to make it the default it has to be called `4byteTracer`

**Comment by holiman** [comment]
After a fix-up, the discrepancies are gone. 
@wardbradt you seem to have disabled maintainer-commit on your fork of geth, so I can't push to this branch. 
Can you cherry-pick this commit https://github.com/holiman/go-ethereum/commit/d57724ba1e6afcda1e2072287ef3148d63c17b64 -- of my `fixup` branch?

**Review by s1na — APPROVED** [review]
Very cool, thanks!

**Comment by holiman** [comment]
Second attempt, 

js : `12m15.214591921s`
native (warmed up by the previous run): `4m30.635180831s`

Yes, I think we should make the go-tracer be the default, so register it as `4byteTracer`

**Review by holiman — APPROVED** [review]

**Review by holiman — APPROVED** [review]
LGTM

**Comment by holiman** [comment]
TODO for follow-up PR: make the golang tracer default

**Pull request merged**